Source/Metal/DeviceMetal.mm

```diff
@@ -0,0 +1,57 @@
+#include "DeviceMetal.h"
+
+using namespace nri;
+
+BufferMetal::BufferMetal(const CallbackInterface& callbacks, const StdAllocator<uint8_t>& stdAllocator) 
+    : DeviceBase(callbacks, stdAllocator) {
+    m_Desc.graphicsAPI = GraphicsAPI::VK;
+    m_Desc.nriVersionMajor = NRI_VERSION_MAJOR;
+    m_Desc.nriVersionMinor = NRI_VERSION_MINOR;
+
+}
+BufferMetal::~BufferMetal() {
+
+}
+
+Result CreateDeviceMTL(const DeviceCreationDesc& deviceCreationDesc, DeviceBase*& device) {
+    
+}
+
+Result CreateDeviceMTL(const DeviceCreationMTLDesc& deviceCreationDesc, DeviceBase*& device) {
+
+}
+
+Result DeviceMetal::Create(const DeviceCreationDesc& deviceCreationDesc, const DeviceCreationMTLDesc& deviceCreationMTLDesc, bool isWrapper) {
+  m_OwnsNativeObjects = !isWrapper;
+
+  if(isWrapper) {
+    m_Device = deviceCreationMTLDesc.MtlDevice;
+  }
+  
+  strncpy(m_Desc.adapterDesc.name, [m_Device.name UTF8String], sizeof(m_Desc.adapterDesc.name));
+  // No vendor id, device id for Apple GPUs
+  if (strstr(m_Desc.adapterDesc.name, "Apple"))
+  {
+      m_Desc.adapterDesc.vendor = nri::Vendor::APPLE;
+  } else {
+      const uint64_t regID = [m_Device registryID];
+      if (regID)
+      {
+          io_registry_entry_t entry = IOServiceGetMatchingService(kIOMasterPortDefault, IORegistryEntryIDMatching(regID));
+          if (entry)
+          {
+              // That returned the IOGraphicsAccelerator nub. Its parent, then, is the actual PCI device.
+              io_registry_entry_t parent;
+              if (IORegistryEntryGetParentEntry(entry, kIOServicePlane, &parent) == kIOReturnSuccess)
+              {
+                  m_Desc.adapterDesc.vendor = GetVendorFromID(GetEntryProperty(parent, CFSTR("vendor-id")));
+                  m_Desc.adapterDesc.deviceId = GetEntryProperty(parent, CFSTR("device-id"));
+                  IOObjectRelease(parent);
+              }
+              IOObjectRelease(entry);
+          }
+      }
+  }
```

Comment:
something interesting, I guess apple doesn't report the device/vendor.
```
    "Apple": {
      "id": "0x106b",

      "_comment": [
        "Apple GPUs do not report a DeviceID via the Metal API, and as such the typical device",
        "pattern matching does not work for them. The recommended approach is to find the highest",
        "supported 'common' family supported and report it as the architecture.",
        "Examples: 'common-1', 'common-3'",
        "https://developer.apple.com/documentation/metal/gpu_devices_and_work_submission/detecting_gpu_features_and_metal_software_versions"
      ]
    },
```



```
  // No vendor id, device id for Apple GPUs
  if (strstr(m_Desc.adapterDesc.name, "Apple")) {
      m_Desc.adapterDesc.vendor = nri::Vendor::APPLE;
  } else {
      const uint64_t regID = [m_Device registryID];
      if (regID)
      {
          io_registry_entry_t entry = IOServiceGetMatchingService(kIOMasterPortDefault, IORegistryEntryIDMatching(regID));
          if (entry)
          {
              // That returned the IOGraphicsAccelerator nub. Its parent, then, is the actual PCI device.
              io_registry_entry_t parent;
              if (IORegistryEntryGetParentEntry(entry, kIOServicePlane, &parent) == kIOReturnSuccess)
              {
                  m_Desc.adapterDesc.vendor = GetVendorFromID(GetEntryProperty(parent, CFSTR("vendor-id")));
                  m_Desc.adapterDesc.deviceId = GetEntryProperty(parent, CFSTR("device-id"));
                  IOObjectRelease(parent);
              }
              IOObjectRelease(entry);
          }
      }
  }
```

Reply:
I don't think it's a problem. Just do what Metal's layer says and just pass the device.